modules/shared/src/main/scala/org/tessellation/currency/utils/DataTransactionUtils.scala

```diff
@@ -0,0 +1,70 @@
+package org.tessellation.currency.utils
+
+import cats.data.NonEmptyList
+import cats.effect.Async
+import cats.syntax.all._
+
+import org.tessellation.currency.dataApplication.FeeTransaction.serializeFeeTransaction
+import org.tessellation.currency.dataApplication.{DataTransaction, DataUpdate, FeeTransaction}
+import org.tessellation.security.hash.Hash
+import org.tessellation.security.signature.Signed
+import org.tessellation.security.{Hasher, hash}
+
+object DataTransactionUtils {
+  private def collectTransactions[A <: DataTransaction](dataTransactions: List[NonEmptyList[Signed[DataTransaction]]])(
+    pf: PartialFunction[DataTransaction, A]
+  ): List[Signed[A]] =
+    dataTransactions.flatMap { currentDataTransactions =>
+      currentDataTransactions.collect { dataTransaction =>
+        dataTransaction.value match {
+          case transaction if pf.isDefinedAt(transaction) =>
+            Signed(pf(transaction), dataTransaction.proofs)
+        }
+      }
+    }
+
+  def getDataUpdates(dataTransactions: List[NonEmptyList[Signed[DataTransaction]]]): List[Signed[DataUpdate]] =
+    collectTransactions(dataTransactions) {
+      case dataUpdate: DataUpdate => dataUpdate
+    }
+
+  def getFeeTransactions(dataTransactions: List[NonEmptyList[Signed[DataTransaction]]]): List[Signed[FeeTransaction]] =
+    collectTransactions(dataTransactions) {
+      case feeTransaction: FeeTransaction => feeTransaction
+    }
+
+  def getDataTransactionHashes[F[_]: Async: Hasher](
+    data: List[NonEmptyList[Signed[DataTransaction]]],
+    serializeDataUpdate: DataUpdate => F[Array[Byte]]
+  ): F[List[NonEmptyList[hash.Hash]]] =
+    data.traverse { dataTransactions =>
+      dataTransactions.toList.traverse { signedTransaction =>
+        signedTransaction.value match {
+          case dataUpdate: DataUpdate =>
+            Signed(dataUpdate, signedTransaction.proofs)
+              .toHashed(serializeDataUpdate)
+              .map(_.hash)
+          case feeTransaction: FeeTransaction =>
+            Signed(feeTransaction, signedTransaction.proofs)
+              .toHashed(serializeFeeTransaction[F])
+              .map(_.hash)
+        }
+      }.map(NonEmptyList.fromListUnsafe)
+    }
+
+  def getFeeTransactionOfDataUpdate[F[_]: Async](
+    dataTransactions: NonEmptyList[Signed[DataTransaction]],
+    dataUpdate: DataUpdate,
+    serializeDataUpdate: DataUpdate => F[Array[Byte]]
+  ): F[Option[Signed[FeeTransaction]]] = {
+    val feeTransactions = dataTransactions.collect {
+      case Signed(feeTransaction: FeeTransaction, proofs) => Signed(feeTransaction, proofs)
+    }
+
+    serializeDataUpdate(dataUpdate).flatMap { serializedDataUpdate =>
+      feeTransactions.findM { feeTransaction =>
+        Async[F].delay(Hash.fromBytes(serializedDataUpdate) === feeTransaction.value.dataUpdateRef)
```

Comment:
equality fn does not need to be delayed. if there is an operation that needs to be delayed, it's better to make it step by step (separate chained operations)

Reply:
👍🏼